xfer/model_handler/model_handler.py

```diff
@@ -43,103 +44,498 @@ def __init__(self, module, context_function=mx.context.cpu, num_devices=1, data_
+                                                                    nodes_before=temp_symbol_dict[consts.NODES],
+                                                                    deleted_node_operator_idx=del_node_op_idx)
+            if join_deleted:
+                layers_dropped.append(join_layer_name)
+
+            # Update symbol dictionary attributes
+            symbol_dict[consts.ARG_NODES] = self._get_arg_nodes(symbol_dict[consts.NODES])
+            symbol_dict[consts.HEADS] = self._get_heads(nodes=symbol_dict[consts.NODES],
+                                                        output_layer_names=self._get_output_layer_names(
+                                                            nodes=temp_symbol_dict[consts.NODES],
+                                                            heads=temp_symbol_dict[consts.HEADS]))
+            symbol_dict[consts.NODES] = self._update_inputs(nodes=symbol_dict[consts.NODES],
+                                                            original_nodes=temp_symbol_dict[consts.NODES],
+                                                            drop_layer_name=drop_layer_name,
+                                                            join_deleted=join_deleted,
+                                                            join_idx=join_idx)
+
+        sym = mx.sym.load_json(json.dumps(symbol_dict))
+
+        logging.info('{} deleted from model bottom'.format(', '.join(layers_dropped)))
+        if drop_layer_names is not None:
+            if len(drop_layer_names) > 0:
+                logging.warning('Did not use all of drop_layer_names: {}'.format(', '.join(drop_layer_names)))
         self.update_sym(sym)
 
-    def add_layer_top(self, layer_factory_list):
+    def add_layer_top(self, layer_list):
         """
         Add layer to output of model.
 
-        :param layer_factory: List of LayerFactory objects to be added to model output.
-        :type layer_factory: list(:class:`LayerFactory`)
+        :param layer_list: List of MxNet symbol layers to be added to model output.
+        :type layer_list: list(:class:`mx.symbol`)
         """
         if '_label' in self.symbol.get_internals().list_outputs()[consts.LABEL_IDX]:
             raise exceptions.ModelError('Cannot add layer above output layer')
-        sym = self.symbol
         added_layer_names = []
-        for lf in layer_factory_list:
-            self._validate_layer_name(lf.attributes[consts.NAME])
-            sym = lf.create_layer(sym)
-            added_layer_names.append(lf.attributes[consts.NAME])
+
+        net_symbol_dict = self._get_symbol_dict(self.symbol)
+        for layer_symbol in layer_list:
+            added_layer_names.append(layer_symbol.name)
+            layer_symbol_dict = self._get_symbol_dict(layer_symbol)
+
+            num_existing_nodes = len(net_symbol_dict[consts.NODES])
+
+            # Update inputs of nodes
```

Comment:
Consider something like:
`# Shift input indices of new layer using index of the last node in existing network.`

Reply:
Updated comment